hnix-store-remote/src/System/Nix/Store/Remote/Client/Core.hs

```diff
@@ -75,14 +81,57 @@ doReq = \case
           Nothing ->
             throwError
               RemoteStoreError_NoNarSourceProvided
+        processOutput
+        processReply
+      AddToStoreNar _ meta _ _ -> do
+        let narBytes = maybe 0 id $ metadataNarBytes meta
+        maybeDataSource <- takeDataSource
+        soc <- getStoreSocket
+        case maybeDataSource of
+          Nothing ->
+            if narBytes == 0 then writeFramedSource (const (pure Nothing)) soc 0
+            else throwError RemoteStoreError_NoDataSourceProvided
+          Just dataSource -> do
+            writeFramedSource dataSource soc narBytes
```

Comment:
Now I kinda wonder what's the `Logger_Read` code path for.

Reply:
Yeah, I don't know if it's still used in the newer protocol versions.